org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java

```diff
@@ -157,28 +216,132 @@ public void accept(ISourceType[] sourceType, PackageBinding packageBinding,
 				// from parsed files
 				requestor.acceptBinding(bindingKey, bindingFromMap);
 			} else {
-				// from ECJ
-				char[] charArrayFQN = Signature.toCharArray(bindingKey.toCharArray());
-				char[][] twoDimensionalCharArrayFQN = Stream.of(new String(charArrayFQN).split("/")) //
-						.map(myString -> myString.toCharArray()) //
-						.toArray(char[][]::new);
-
-				NameEnvironmentAnswer answer = environment.findType(twoDimensionalCharArrayFQN);
-				IBinaryType binaryType = answer.getBinaryType();
-				if (binaryType != null) {
-					BinaryTypeBinding binding = lu.cacheBinaryType(binaryType, null);
-					requestor.acceptBinding(bindingKey, new TypeBinding(bindingResolver, binding));
+				
+				CustomBindingKeyParser bkp = new CustomBindingKeyParser(bindingKey);
+				bkp.parse(true);
+				char[][] name = bkp.compoundName;
+
+//				// from ECJ
+//				char[] charArrayFQN = Signature.toCharArray(bindingKey.toCharArray());
+//				char[][] twoDimensionalCharArrayFQN = Stream.of(new String(charArrayFQN).split("/")) //
+//						.map(myString -> myString.toCharArray()) //
+//						.toArray(char[][]::new);
+//				char[][] twoDimensionalCharArrayFQN = new char[][] {};
+				NameEnvironmentAnswer answer = environment.findType(name);
+				if( answer != null ) {
+					IBinaryType binaryType = answer.getBinaryType();
+					if (binaryType != null) {
+						BinaryTypeBinding binding = lu.cacheBinaryType(binaryType, null);
+						requestor.acceptBinding(bindingKey, new TypeBinding(bindingResolver, binding));
+					}
 				}
 			}
 
 		}
 
 	}
 
+	private static class CustomBindingKeyParser extends BindingKeyParser {
+
+		private char[] secondarySimpleName;
+		private char[][] compoundName;
+
+		public CustomBindingKeyParser(String key) {
+			super(key);
+		}
+
+		@Override
+		public void consumeSecondaryType(char[] simpleTypeName) {
+			this.secondarySimpleName = simpleTypeName;
+		}
+
+		@Override
+		public void consumeFullyQualifiedName(char[] fullyQualifiedName) {
+			this.compoundName = CharOperation.splitOn('/', fullyQualifiedName);
+		}
+	}
+	
+	@Override
+	public IBinding[] resolve(IJavaElement[] elements, int apiLevel, Map<String, String> compilerOptions,
+			IJavaProject javaProject, WorkingCopyOwner owner, int flags, IProgressMonitor monitor) {
+		// Logic copied entirely from ECJCompilationUnitResolver
```

Comment:
Do you think this could be factorized somehow, for example by having it default implementation in ICompilationUnitResolver ?